src/hardware/pcspeaker_discrete.cpp

```diff
@@ -0,0 +1,491 @@
+			pit_last = 0;
+			pit_mode = PitMode::InterruptOnTC;
+			return;
+		}
+		pit_new_max  = PERIOD_OF_1K_PIT_TICKS * cntr;
+		pit_new_half = pit_new_max / 2;
+		break;
+	case PitMode::SoftwareStrobe: /* Software triggered strobe */
+		pit_last = amp_positive;
+		AddDelayEntry(newindex, pit_last);
+		pit_index = 0;
+		pit_max   = PERIOD_OF_1K_PIT_TICKS * cntr;
+		break;
+	default:
+#if C_DEBUG
+		LOG_WARNING("PCSPEAKER: Unhandled speaker PIT mode: %s", PitModeToString(pit_mode));
+#endif
+		return;
+	}
+	// Activate the channel after queuing new speaker entries
+	channel->Enable(true);
+}
+
+// Returns the amp_neutral voltage if the speaker's  fully faded,
+// otherwise returns the fallback if the speaker is active.
+double PcSpeakerDiscrete::NeutralOr(double fallback)
+{
+	return !idle_countdown ? amp_neutral : fallback;
+}
+
+// Returns, in order of preference:
+// - Neutral voltage, if the speaker's fully faded
+// - The last active PIT voltage to stitch on-going playback
+// - The fallback voltage to kick start a new sound pattern
+double PcSpeakerDiscrete::NeutralLastPitOr(double fallback)
+{
+	const bool use_last = std::isgreater(fabs(pit_last), amp_neutral);
+	return NeutralOr(use_last ? pit_last : fallback);
+}
+
+// PWM-mode activation
+void PcSpeakerDiscrete::SetType(const PpiPortB &b)
+{
+	const auto newindex = PIC_TickIndex();
+	ForwardPIT(newindex);
+
+	prev_port_b_state.data = port_b_state.data;
+
+	switch (b.timerGateAndSpeakerOutput) {
+	case 0b00:
+		port_b_state.timerGateOutput = false;
+		port_b_state.speakerOutput   = false;
+		AddDelayEntry(newindex, NeutralOr(amp_negative));
+		break;
+	case 0b01:
+		port_b_state.timerGateOutput = false;
+		port_b_state.speakerOutput   = true;
+		AddDelayEntry(newindex, NeutralOr(amp_negative));
+		break;
+	case 0b10:
+		port_b_state.timerGateOutput = true;
+		port_b_state.speakerOutput   = false;
+		AddDelayEntry(newindex, NeutralOr(amp_positive));
+		break;
+	case 0b11:
+		port_b_state.timerGateOutput = true;
+		port_b_state.speakerOutput   = true;
+		AddDelayEntry(newindex, NeutralLastPitOr(amp_positive));
+		break;
+	};
+
+	// Activate the channel after queuing new speaker entries
+	channel->Enable(true);
+}
+
+// After the speaker has idled (as defined when the number of speaker
+// movements is zero) for a short time, wind down the DC-offset, and
+// then halt the channel.
+void PcSpeakerDiscrete::PlayOrFadeout(const uint16_t speaker_movements,
+                                      size_t requested_samples, int16_t *buffer)
+{
+	if (speaker_movements && requested_samples) {
+		static_assert(idle_grace_time_ms > 0, "Algorithm depends on a non-zero grace time");
+		idle_countdown = idle_grace_time_ms;
+		dc_silencer.Reset();
+	} else if (idle_countdown > 0) {
+		idle_countdown--;
+		last_played_sample = buffer[requested_samples - 1];
+	} else if (!dc_silencer.Generate(last_played_sample, requested_samples, buffer)) {
+		channel->Enable(false);
+	}
```

Comment:
Strictly speaking, we don't need to DC silencer at all with the master high-pass filter in place. I was meant to remove it altogether because I don't think it's necessary anywhere now, just forgot.

Reply:
Removed!